### DISCUSSION TITLE
Is there a way to inline enums?

### DISCUSSION THREAD

**Discussion by electrosaur** [victools/jsonschema-generator#176] [Q&A] [answered]
If I have `X{TimeUnit y}` I'd like to produce `{"X": "object {"y": "string" enum ["HOURS", "DAYS", ...]}}`.

**Comment by CarstenWickner** [comment, marked as answer]
Hi @electrosaur,

Is the `Option.FLATTENED_ENUMS` fulfilling your requirement?

As per https://victools.github.io/jsonschema-generator/#generator-options
> #### 5 - `Option.FLATTENED_ENUMS`
> Behaviour if included:
> - Treating enountered enum types as `{ "type": "string" }` schema with the names of the enum constants being listed as its `enum` values.

That should produce something like this:
```json
{
  "type": "object",
  "properties": {
    "y": {
      "type": "string",
      "enum": ["HOURS", "DAYS", ...]
    }
  }
}
```

----

If you want to generally avoid that enums are being referenced via the `$defs`/`definitions`, then please refer to my answer on this issue: https://github.com/victools/jsonschema-generator/issues/123
> to inline all enums you can use `configBuilder.forTypesInGeneral().withCustomDefinitionProvider(new InlineAllEnumsDefinitionProvider())` with an instance of this custom definition provider:
> ```java
> class InlineAllEnumsDefinitionProvider implements CustomDefinitionProviderV2 {
>     @Override
>     public CustomDefinition provideCustomSchemaDefinition(ResolvedType javaType, SchemaGenerationContext context) {
>         if (javaType.isInstanceOf(Enum.class)) {
>             ObjectNode standardDefinition = context.createStandardDefinition(javaType, this);
>             return new CustomDefinition(standardDefinition, true);
>         }
>         return null;
>     }
> }
> ```

The documentation of custom definition providers in general can be found here:
https://victools.github.io/jsonschema-generator/#custom-type-definitions

**Comment by electrosaur** [comment]
Hi Carsten:

Thanks for the prompt response.

Maybe I'm not doing it right. I have say TimeUnit timeUnit as a member 
of my POJO. I'm getting

`timeUnit -> "{"$ref":"#/$defs/java.util.concurrent.TimeUnit"}"`

Maybe it's clashing with the DEFINITIONS_FOR_ALL_OBJECTS?

My code is:

```java
SchemaGeneratorConfigBuilder configBuilder =new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2019_09,OptionPreset.PLAIN_JSON)
     .with(module)
     .with(FLATTENED_ENUMS)
     .with(DEFINITIONS_FOR_ALL_OBJECTS);
configBuilder.forTypesInGeneral()
     .withIdResolver(scope ->scope.getType().getErasedType().getName());
configBuilder.forTypesInGeneral()
     .withDefinitionNamingStrategy(new DefaultSchemaDefinitionNamingStrategy() {
         @Override public String getDefinitionNameForKey(DefinitionKey key,SchemaGenerationContext generationContext) {
             return key.getType().getErasedType().getName();
         }
     });
```
And I'm reading the fields of the object via

`schema.get("properties").fields().forEachRemaining(...)`

Shankar

**Reply by CarstenWickner** [reply]
Hi Shankar,

Indeed, the `Option.DEFINITIONS_FOR_ALL_OBJECTS` will result in your enum being moved into the `$defs`/`definitions` together with all other sub-schemas.

_EDIT: I'll include the particular solution in my answer above._
